File: 2023年/5月/5号/95.不同的二叉搜索树-ii.cpp
/*
 * @lc app=leetcode.cn id=95 lang=cpp
 *
 * [95] 不同的二叉搜索树 II
 */

#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// @lc code=start
/**
Accepted
8/8 cases passed (16 ms)
Your runtime beats 77.29 % of cpp submissions
Your memory usage beats 32.62 % of cpp submissions (15.8 MB)
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
    vector<TreeNode *> generateTrees(int start, int end)
    {
        if (start > end)
        {
            return {nullptr};
        }
        vector<TreeNode *> root;
        for (int i = start; i <= end; i++)
        {
            vector<TreeNode *> left = generateTrees(start, i - 1);
            vector<TreeNode *> right = generateTrees(i + 1, end);
            for (TreeNode *left : left)
            {
                for (TreeNode *right : right)
                {
                    TreeNode *node = new TreeNode(i);
                    node->left = left;
                    node->right = right;
                    root.push_back(node);
                }
            }
        }
        return root;
    }

public:
    vector<TreeNode *> generateTrees(int n)
    {
        if (n == 0)
        {
            return {};
        }
        return generateTrees(1, n);
    }
};
// @lc code=end

File: 2025年/1月/16号/95.不同的二叉搜索树-ii.cs
/*
 * @lc app=leetcode.cn id=95 lang=csharp
 *
 * [95] 不同的二叉搜索树 II
 */
public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    //递归
    public List<TreeNode> GenerateTrees(int start, int end)
    {
        List<TreeNode> result = new List<TreeNode>();
        //递归出口
        if (start > end)
        {
            result.Add(null);
            return result;
        }
        for (int i = start; i <= end; i++)
        {
            //递归
            List<TreeNode> lefts = GenerateTrees(start, i - 1);
            List<TreeNode> rights = GenerateTrees(i + 1, end);
            //穷举所有可能的左右子树
            foreach (var left in lefts)
            {
                foreach (var right in rights)
                {
                    TreeNode root = new TreeNode(i, left, right);
                    root.left = left;
                    root.right = right;
                    result.Add(root);
                }
            }
        }
        return result;
    }
    public IList<TreeNode> GenerateTrees(int n)
    {
        if (n == 0) { return new List<TreeNode>(); }
        return GenerateTrees(1, n);
    }
}
// @lc code=end

